cpp/src/arrow/scalar.cc

```diff
@@ -153,9 +153,10 @@ struct ScalarHashImpl {
 
   Status ArrayHash(const ArrayData& a) {
     RETURN_NOT_OK(StdHash(a.length) & StdHash(a.GetNullCount()));
-    if (a.buffers[0] != nullptr) {
+    if (a.GetNullCount() != 0) {
```

Comment:
IIUC, the null count should get cached from calling `GetNullCount()` above, so this line does not have performance degradations.

Reply:
Indeed, however, a NullArray can have a non-zero null count but without a null bitmap. So you must make sure to check the buffers as well.